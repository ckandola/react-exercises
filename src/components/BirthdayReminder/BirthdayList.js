import React from 'react';
import dateFormat from 'dateformat';
import PropTypes from 'prop-types';

const BirthdayList = ({currentDay, people, onCongratulate}) => {

    const filterList = list => {
        return list.filter(person => {
            return dateFormat(person.birthday, 'mmdd') === dateFormat(currentDay, 'mmdd');
        })
    }

    const filteredList = filterList(people);

    return (
        <>
            <h3>{`${filteredList.length} birthday${filteredList.length !== 1 ? 's' : ''} today`}</h3>
            {filteredList.map(person => {
                const {
                    id, 
                    birthday, 
                    age = Number(dateFormat(currentDay), 'yyyy') - Number(dateFormat(birthday, 'yyyy')),
                    name, 
                    image
                } = person;

                return (
                    <article key={id} className="birthday-person">
                        <img src={image} alt={name}/>
                        <div>
                            <h4>{name}</h4>
                            <p>{`${age} years`}</p>
                        </div>
                        <button className="birthday-congrat-btn" onClick={() => onCongratulate(id)}>Congratulate</button>
                    </article>
                );
            })}
        </>
    );
};

export default BirthdayList;

BirthdayList.propTypes = {
    currentDay: PropTypes.string,
    people: PropTypes.arrayOf(PropTypes.shape({
        id: PropTypes.number.isRequired,
        birthday: PropTypes.string.isRequired,
        name: PropTypes.string.isRequired,
        age: PropTypes.number,
        image: PropTypes.string
    })),
    onCongratulate: PropTypes.func
};
