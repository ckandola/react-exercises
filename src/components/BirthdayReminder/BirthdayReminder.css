/*
=============== 
Variables
===============
*/

:root {
    --clr-white: #fff;
    --transition: all 0.3s linear;
    --spacing: 0.1rem;
    --radius: 0.25rem;
    --light-shadow: 0 5px 15px rgba(0, 0, 0, 0.1);
    --dark-shadow: 0 5px 15px rgba(0, 0, 0, 0.4);
    --max-width: 1170px;
    --fixed-width: 450px;
  }
  
  /* section */
  .birthday-section {
    width: 90vw;
    margin: 0 auto;
    max-width: var(--max-width);
  }
  
  @media screen and (min-width: 992px) {
    .birthday-section {
      width: 95vw;
    }
  }
  
  .birthday-main {
    min-height: 100vh;
    display: flex;
    justify-content: center;
    align-items: center;
  }
  
  .birthday-container {
    width: 90vw;
    margin: 5rem 0;
    max-width: var(--fixed-width);
    background: var(--clr-white);
    border-radius: var(--radius);
    padding: 1.5rem 2rem;
    box-shadow: var(--dark-shadow);
  }
  .birthday-container h3 {
    font-weight: normal;
    text-transform: none;
    margin-bottom: 1.25rem;
  }
  .birthday-person {
    display: grid;
    grid-template-columns: auto 1fr 2fr;
    column-gap: 0.75rem;
    margin-bottom: 1.5rem;
    align-items: center;
  }
  .birthday-person img {
    width: 75px;
    height: 75px;
    object-fit: cover;
    border-radius: 50%;
    box-shadow: var(--light-shadow);
  }
  .birthday-person h4 {
    margin-bottom: 0.35rem;
  }
  .birthday-person p {
    margin-bottom: 0;
  }
  .birthday-btn {
    color: var(--clr-white);
    display: block;
    width: 100%;
    border-color: transparent;
    background: navy;
    margin: 2rem auto 0 auto;
    text-transform: capitalize;
    font-size: 1.2rem;
    padding: 0.5rem 0;
    letter-spacing: var(--spacing);
    border-radius: var(--radius);
    outline: 1px solid rgba(242, 138, 178, 0.8);
    cursor: pointer;
  }
  
  .birthday-congrat-btn {
    text-align: center;
    padding: 0.5rem;
    border-radius: var(--radius);
    background-color:rgba(242, 138, 178, 0.8);
    width: 50%;
  }
  
  .birthday-input-section {
    text-align: center;
  }
  
  .birthday-input-section-container {
    display:flex;
    height: 3rem;
    align-items: center;
    justify-content: center;
  }
  
  .birthday-reset-sectn {
    position: relative;
    width:9rem;
  }
  
  .birthday-reset-date-btn {
    border-radius: var(--radius);
    width:fit-content;
    position: absolute;
    top: 50%;
    transform: translate(-50%, -50%);
    background-color: skyblue;
  }
