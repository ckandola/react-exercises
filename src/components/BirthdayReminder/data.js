export default [
    {
      id: 1,
      birthday: "06/28/1993",
      name: 'Bertie Yates',
      age: 29,
      image:
        'https://res.cloudinary.com/diqqf3eq2/image/upload/v1595959131/person-2_ipcjws.jpg',
    },
    {
      id: 2,
      birthday: "06/29/1993",
      name: 'Hester Hogan',
      age: 32,
      image:
        'https://res.cloudinary.com/diqqf3eq2/image/upload/v1595959131/person-3_rxtqvi.jpg',
    },
    {
      id: 3,
      birthday: "06/28/1993",
      name: 'Larry Little',
      age: 36,
      image:
        'https://res.cloudinary.com/diqqf3eq2/image/upload/v1586883423/person-4_t9nxjt.jpg',
    },
    {
      id: 4,
      birthday: "06/28/1993",
      name: 'Sean Walsh',
      age: 34,
      image:
        'https://res.cloudinary.com/diqqf3eq2/image/upload/v1586883417/person-3_ipa0mj.jpg',
    },
    {
      id: 5,
      birthday: "01/02/1993",
      name: 'Lola Gardner',
      age: 29,
      image:
        'https://res.cloudinary.com/diqqf3eq2/image/upload/v1586883334/person-1_rfzshl.jpg',
    },
  ];
  