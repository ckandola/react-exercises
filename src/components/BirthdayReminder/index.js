// from https://www.youtube.com/watch?v=a_7Z7C_JCyo

import React, {useState} from 'react';
import data from './data';
import BirthdayList from './BirthdayList';
import './index.css';

const BirthdayReminder = () => {
    const [people, setPeople] = useState(data);
    return (
        <main>
            <section className="container">
                <h3>{`${people.length} birthdays today`}</h3>
                <BirthdayList people={people}/>
                <button onClick={() => setPeople([])}>Clear All</button>
            </section>
        </main>
    );
}

export default BirthdayReminder;
