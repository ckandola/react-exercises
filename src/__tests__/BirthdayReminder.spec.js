import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import BirthdayReminder from '../components/BirthdayReminder';
import BirthdayList from '../components/BirthdayReminder/BirthdayList';

describe('BirthdayReminder tests', () => {
    
    it('contains the correct initial list of birthdays', () => {
        const component = renderer.create(<BirthdayReminder date={new Date('June 28 2021')} />);
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('clears the list of birthdays when clicked', () => {
        const wrapper = shallow(<BirthdayReminder date={new Date('June 28 2021')} />);
        const clearButton = wrapper.find('#birthday-list-clear');
        clearButton.simulate('click');
        expect(wrapper.find(BirthdayList).getElement().props.people).toHaveLength(0);
    });
    
});
