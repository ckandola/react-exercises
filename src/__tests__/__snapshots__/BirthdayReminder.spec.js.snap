// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BirthdayReminder tests contains the correct initial list of birthdays 1`] = `
<main
  className="birthday-main"
>
  <section
    className="birthday-container birthday-section"
  >
    <div
      className="birthday-input-section"
    >
      <h3>
        Search for a birthday:
      </h3>
      <div
        className="birthday-input-section-container"
      >
        <div
          className="react-datepicker-wrapper"
        >
          <div
            className="react-datepicker__input-container"
          >
            <input
              className=""
              disabled={false}
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              placeholder="Enter a day!"
              readOnly={false}
              type="text"
              value="06/28/2021"
            />
          </div>
        </div>
        <div
          className="birthday-reset-sectn"
        >
          <button
            className="birthday-reset-date-btn"
            onClick={[Function]}
          >
            Return to today
          </button>
        </div>
      </div>
    </div>
    <h3>
      3 birthdays today
    </h3>
    <article
      className="birthday-person"
    >
      <img
        alt="Bertie Yates"
        src="https://res.cloudinary.com/diqqf3eq2/image/upload/v1595959131/person-2_ipcjws.jpg"
      />
      <div>
        <h4>
          Bertie Yates
        </h4>
        <p>
          29 years
        </p>
      </div>
      <button
        className="birthday-congrat-btn"
        onClick={[Function]}
      >
        Congratulate
      </button>
    </article>
    <article
      className="birthday-person"
    >
      <img
        alt="Larry Little"
        src="https://res.cloudinary.com/diqqf3eq2/image/upload/v1586883423/person-4_t9nxjt.jpg"
      />
      <div>
        <h4>
          Larry Little
        </h4>
        <p>
          36 years
        </p>
      </div>
      <button
        className="birthday-congrat-btn"
        onClick={[Function]}
      >
        Congratulate
      </button>
    </article>
    <article
      className="birthday-person"
    >
      <img
        alt="Sean Walsh"
        src="https://res.cloudinary.com/diqqf3eq2/image/upload/v1586883417/person-3_ipa0mj.jpg"
      />
      <div>
        <h4>
          Sean Walsh
        </h4>
        <p>
          34 years
        </p>
      </div>
      <button
        className="birthday-congrat-btn"
        onClick={[Function]}
      >
        Congratulate
      </button>
    </article>
    <button
      className="birthday-btn"
      id="birthday-list-clear"
      onClick={[Function]}
    >
      Clear All
    </button>
    <button
      className="birthday-btn"
      onClick={[Function]}
    >
      Reload
    </button>
  </section>
</main>
`;
